TestNoLeak is flaky

```
WARNING: DATA RACE
Read at 0x00c002180528 by goroutine 2594:
  google.golang.org/protobuf/internal/impl.sizeBytesNoZero()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/codec_gen.go:5468 +0x44
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:111 +0x500
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.sizeMessageInfo()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/codec_field.go:160 +0x80
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:111 +0x500
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.sizeMessageInfo()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/codec_field.go:160 +0x80
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:111 +0x500
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.(*MessageInfo).size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:41 +0x88
  google.golang.org/protobuf/internal/impl.(*MessageInfo).size-fm()
      <autogenerated>:1 +0x54
  google.golang.org/protobuf/proto.MarshalOptions.size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:43 +0x1a0
  google.golang.org/protobuf/proto.MarshalOptions.Size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:34 +0x5c
  google.golang.org/protobuf/proto.Size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:20 +0x78
  google.golang.org/grpc/encoding/proto.(*codecV2).Marshal()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/encoding/proto/proto.go:51 +0x60
  google.golang.org/grpc.encode()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/rpc_util.go:722 +0x50
  google.golang.org/grpc.prepareMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1862 +0x130
  google.golang.org/grpc.(*serverStream).SendMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1710 +0x2b8
  cloud.google.com/go/spanner/apiv1/spannerpb.(*spannerExecuteStreamingSqlServer).Send()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:818 +0x54
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).executeStreamingSQL()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:967 +0xa28
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).ExecuteStreamingSql()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:910 +0x9c
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:805 +0x144
  google.golang.org/grpc.(*serverStream).RecvMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1791 +0x234
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:802 +0x5c
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1722 +0x178c
  google.golang.org/grpc.(*Server).handleStream()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1846 +0xe14
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0xe8

Previous write at 0x00c002180528 by goroutine 2629:
  github.com/googleapis/go-sql-spanner/testutil.StatementResult.getResultSetWithTransactionSet()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:262 +0x51c
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).executeStreamingSQL()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:939 +0x3cc
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).ExecuteStreamingSql()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:910 +0x9c
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:805 +0x144
  google.golang.org/grpc.(*serverStream).RecvMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1791 +0x234
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:802 +0x5c
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1722 +0x178c
  google.golang.org/grpc.(*Server).handleStream()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1846 +0xe14
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0xe8

Goroutine 2594 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x1e4
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:663 +0x2efc
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:704 +0x2a0
  google.golang.org/grpc.(*Server).serveStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1055 +0x4d0
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:989 +0x7c

Goroutine 2629 (finished) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x1e4
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:663 +0x2efc
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:704 +0x2a0
  google.golang.org/grpc.(*Server).serveStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1055 +0x4d0
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:989 +0x7c
==================
==================
WARNING: DATA RACE
Read at 0x00c002180540 by goroutine 2594:
  google.golang.org/protobuf/internal/impl.pointer.Elem()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/pointer_unsafe.go:124 +0x4b4
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:108 +0x480
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.sizeMessageInfo()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/codec_field.go:160 +0x80
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:111 +0x500
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.sizeMessageInfo()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/codec_field.go:160 +0x80
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:111 +0x500
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.(*MessageInfo).size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:41 +0x88
  google.golang.org/protobuf/internal/impl.(*MessageInfo).size-fm()
      <autogenerated>:1 +0x54
  google.golang.org/protobuf/proto.MarshalOptions.size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:43 +0x1a0
  google.golang.org/protobuf/proto.MarshalOptions.Size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:34 +0x5c
  google.golang.org/protobuf/proto.Size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:20 +0x78
  google.golang.org/grpc/encoding/proto.(*codecV2).Marshal()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/encoding/proto/proto.go:51 +0x60
  google.golang.org/grpc.encode()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/rpc_util.go:722 +0x50
  google.golang.org/grpc.prepareMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1862 +0x130
  google.golang.org/grpc.(*serverStream).SendMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1710 +0x2b8
  cloud.google.com/go/spanner/apiv1/spannerpb.(*spannerExecuteStreamingSqlServer).Send()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:818 +0x54
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).executeStreamingSQL()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:967 +0xa28
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).ExecuteStreamingSql()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:910 +0x9c
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:805 +0x144
  google.golang.org/grpc.(*serverStream).RecvMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1791 +0x234
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:802 +0x5c
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1722 +0x178c
  google.golang.org/grpc.(*Server).handleStream()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1846 +0xe14
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0xe8

Previous write at 0x00c002180540 by goroutine 2629:
  github.com/googleapis/go-sql-spanner/testutil.StatementResult.getResultSetWithTransactionSet()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:262 +0x51c
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).executeStreamingSQL()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:939 +0x3cc
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).ExecuteStreamingSql()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:910 +0x9c
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:805 +0x144
  google.golang.org/grpc.(*serverStream).RecvMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1791 +0x234
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:802 +0x5c
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1722 +0x178c
  google.golang.org/grpc.(*Server).handleStream()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1846 +0xe14
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0xe8

Goroutine 2594 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x1e4
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:663 +0x2efc
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:704 +0x2a0
  google.golang.org/grpc.(*Server).serveStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1055 +0x4d0
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:989 +0x7c

Goroutine 2629 (finished) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x1e4
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:663 +0x2efc
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:704 +0x2a0
  google.golang.org/grpc.(*Server).serveStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1055 +0x4d0
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:989 +0x7c
==================
==================
WARNING: DATA RACE
Read at 0x00c002180510 by goroutine 2594:
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:115 +0x658
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.sizeMessageInfo()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/codec_field.go:160 +0x80
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:111 +0x500
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.sizeMessageInfo()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/codec_field.go:160 +0x80
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:111 +0x500
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.(*MessageInfo).size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:41 +0x88
  google.golang.org/protobuf/internal/impl.(*MessageInfo).size-fm()
      <autogenerated>:1 +0x54
  google.golang.org/protobuf/proto.MarshalOptions.size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:43 +0x1a0
  google.golang.org/protobuf/proto.MarshalOptions.Size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:34 +0x5c
  google.golang.org/protobuf/proto.Size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:20 +0x78
  google.golang.org/grpc/encoding/proto.(*codecV2).Marshal()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/encoding/proto/proto.go:51 +0x60
  google.golang.org/grpc.encode()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/rpc_util.go:722 +0x50
  google.golang.org/grpc.prepareMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1862 +0x130
  google.golang.org/grpc.(*serverStream).SendMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1710 +0x2b8
  cloud.google.com/go/spanner/apiv1/spannerpb.(*spannerExecuteStreamingSqlServer).Send()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:818 +0x54
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).executeStreamingSQL()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:967 +0xa28
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).ExecuteStreamingSql()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:910 +0x9c
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:805 +0x144
  google.golang.org/grpc.(*serverStream).RecvMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1791 +0x234
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:802 +0x5c
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1722 +0x178c
  google.golang.org/grpc.(*Server).handleStream()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1846 +0xe14
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0xe8

Previous write at 0x00c002180510 by goroutine 2629:
  github.com/googleapis/go-sql-spanner/testutil.StatementResult.getResultSetWithTransactionSet()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:262 +0x51c
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).executeStreamingSQL()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:939 +0x3cc
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).ExecuteStreamingSql()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:910 +0x9c
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:805 +0x144
  google.golang.org/grpc.(*serverStream).RecvMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1791 +0x234
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:802 +0x5c
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1722 +0x178c
  google.golang.org/grpc.(*Server).handleStream()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1846 +0xe14
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0xe8

Goroutine 2594 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x1e4
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:663 +0x2efc
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:704 +0x2a0
  google.golang.org/grpc.(*Server).serveStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1055 +0x4d0
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:989 +0x7c

Goroutine 2629 (finished) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x1e4
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:663 +0x2efc
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:704 +0x2a0
  google.golang.org/grpc.(*Server).serveStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1055 +0x4d0
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:989 +0x7c
==================
==================
WARNING: DATA RACE
Write at 0x00c002180508 by goroutine 2594:
  ??()
      -:0 +0x10251df60
  sync/atomic.StoreInt32()
      <autogenerated>:1 +0x14
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.sizeMessageInfo()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/codec_field.go:160 +0x80
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:111 +0x500
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.sizeMessageInfo()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/codec_field.go:160 +0x80
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointerSlow()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:111 +0x500
  google.golang.org/protobuf/internal/impl.(*MessageInfo).sizePointer()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:60 +0x11c
  google.golang.org/protobuf/internal/impl.(*MessageInfo).size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/internal/impl/encode.go:41 +0x88
  google.golang.org/protobuf/internal/impl.(*MessageInfo).size-fm()
      <autogenerated>:1 +0x54
  google.golang.org/protobuf/proto.MarshalOptions.size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:43 +0x1a0
  google.golang.org/protobuf/proto.MarshalOptions.Size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:34 +0x5c
  google.golang.org/protobuf/proto.Size()
      /Users/runner/go/pkg/mod/google.golang.org/protobuf@v1.36.10/proto/size.go:20 +0x78
  google.golang.org/grpc/encoding/proto.(*codecV2).Marshal()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/encoding/proto/proto.go:51 +0x60
  google.golang.org/grpc.encode()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/rpc_util.go:722 +0x50
  google.golang.org/grpc.prepareMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1862 +0x130
  google.golang.org/grpc.(*serverStream).SendMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1710 +0x2b8
  cloud.google.com/go/spanner/apiv1/spannerpb.(*spannerExecuteStreamingSqlServer).Send()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:818 +0x54
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).executeStreamingSQL()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:967 +0xa28
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).ExecuteStreamingSql()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:910 +0x9c
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:805 +0x144
  google.golang.org/grpc.(*serverStream).RecvMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1791 +0x234
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:802 +0x5c
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1722 +0x178c
  google.golang.org/grpc.(*Server).handleStream()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1846 +0xe14
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0xe8

Previous write at 0x00c002180508 by goroutine 2629:
  github.com/googleapis/go-sql-spanner/testutil.StatementResult.getResultSetWithTransactionSet()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:262 +0x51c
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).executeStreamingSQL()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:939 +0x3cc
  github.com/googleapis/go-sql-spanner/testutil.(*inMemSpannerServer).ExecuteStreamingSql()
      /Users/runner/work/go-sql-spanner/go-sql-spanner/testutil/inmem_spanner_server.go:910 +0x9c
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:805 +0x144
  google.golang.org/grpc.(*serverStream).RecvMsg()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/stream.go:1791 +0x234
  cloud.google.com/go/spanner/apiv1/spannerpb._Spanner_ExecuteStreamingSql_Handler()
      /Users/runner/go/pkg/mod/cloud.google.com/go/spanner@v1.86.1/apiv1/spannerpb/spanner_grpc.pb.go:802 +0x5c
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1722 +0x178c
  google.golang.org/grpc.(*Server).handleStream()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1846 +0xe14
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1061 +0xe8

Goroutine 2594 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x1e4
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:663 +0x2efc
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:704 +0x2a0
  google.golang.org/grpc.(*Server).serveStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1055 +0x4d0
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:989 +0x7c

Goroutine 2629 (finished) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1072 +0x1e4
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:663 +0x2efc
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/internal/transport/http2_server.go:704 +0x2a0
  google.golang.org/grpc.(*Server).serveStreams()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:1055 +0x4d0
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /Users/runner/go/pkg/mod/google.golang.org/grpc@v1.76.0/server.go:989 +0x7c
==================
--- FAIL: TestNoLeak (1.43s)
    testing.go:1490: race detected during execution of test
```

--- a/aborted_transactions_test.go
+++ b/aborted_transactions_test.go
@@ -442,7 +442,7 @@ func testRetryReadWriteTransactionWithQuery(t *testing.T, setupServer func(serve
 	if setupServer != nil {
 		setupServer(server.TestSpanner)
 	}
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 1000*time.Second)
 	defer cancel()
 	tx, err := db.BeginTx(ctx, &sql.TxOptions{})
 	if err != nil {
@@ -470,16 +470,16 @@ func testRetryReadWriteTransactionWithQuery(t *testing.T, setupServer func(serve
 	}
 	err = rows.Err()
 	if g, w := spanner.ErrCode(err), wantErrCode; g != w {
-		t.Fatalf("next error mismatch\nGot: %v\nWant: %v", g, w)
+		t.Fatalf("next error mismatch\n Got: %v\nWant: %v", g, w)
 	}
 	if wantErrCode == codes.OK {
 		if numRowsToConsume > -1 {
 			if g, w := len(values), firstNonZero(numRowsToConsume, 2); g != w {
-				t.Fatalf("row count mismatch\nGot: %v\nWant: %v", g, w)
+				t.Fatalf("row count mismatch\n Got: %v\nWant: %v", g, w)
 			}
 			wantValues := ([]int64{1, 2})[:firstNonZero(numRowsToConsume, 2)]
 			if !cmp.Equal(wantValues, values) {
-				t.Fatalf("values mismatch\nGot: %v\nWant: %v", values, wantValues)
+				t.Fatalf("values mismatch\n Got: %v\nWant: %v", values, wantValues)
 			}
 		}
 	}
@@ -492,16 +492,16 @@ func testRetryReadWriteTransactionWithQuery(t *testing.T, setupServer func(serve
 	}
 	err = tx.Commit()
 	if err != wantCommitErr {
-		t.Fatalf("commit error mismatch\nGot: %v\nWant: %v", err, wantCommitErr)
+		t.Fatalf("commit error mismatch\n Got: %v\nWant: %v", err, wantCommitErr)
 	}
 	reqs := server.TestSpanner.DrainRequestsFromServer()
 	execReqs := testutil.RequestsOfType(reqs, reflect.TypeOf(&sppb.ExecuteSqlRequest{}))
 	if g, w := len(execReqs), wantSqlExecuteCount; g != w {
-		t.Fatalf("execute request count mismatch\nGot: %v\nWant: %v", g, w)
+		t.Fatalf("execute request count mismatch\n Got: %v\nWant: %v", g, w)
 	}
 	commitReqs := testutil.RequestsOfType(reqs, reflect.TypeOf(&sppb.CommitRequest{}))
 	if g, w := len(commitReqs), wantCommitCount; g != w {
-		t.Fatalf("commit request count mismatch\nGot: %v\nWant: %v", g, w)
+		t.Fatalf("commit request count mismatch\n Got: %v\nWant: %v", g, w)
 	}
 
 	// Execute another statement to ensure that the session that was used

--- a/testutil/inmem_spanner_server.go
+++ b/testutil/inmem_spanner_server.go
@@ -135,7 +135,7 @@ type PartialResultSetExecutionTime struct {
 // ToPartialResultSets converts a ResultSet to a PartialResultSet. This method
 // is used to convert a mocked result to a PartialResultSet when one of the
 // streaming methods are called.
-func (s *StatementResult) ToPartialResultSets(resumeToken []byte) (result []*spannerpb.PartialResultSet, err error) {
+func (s *StatementResult) ToPartialResultSets(resumeToken []byte, metadata *spannerpb.ResultSetMetadata) (result []*spannerpb.PartialResultSet, err error) {
 	var startIndex uint64
 	if len(resumeToken) > 0 {
 		if startIndex, err = DecodeResumeToken(resumeToken); err != nil {
@@ -164,9 +164,12 @@ func (s *StatementResult) ToPartialResultSets(resumeToken []byte) (result []*spa
 				rt = s.ResumeTokens[startIndex]
 			}
 			partial := &spannerpb.PartialResultSet{
-				Metadata:    s.ResultSet.Metadata,
 				Values:      values,
 				ResumeToken: rt,
+			}
+			// Only include the metadata in the first PartialResultSet
+			if len(result) == 0 {
+				partial.Metadata = metadata
 			}
 			result = append(result, partial)
 
@@ -179,7 +182,7 @@ func (s *StatementResult) ToPartialResultSets(resumeToken []byte) (result []*spa
 	} else {
 		result = append(result, &spannerpb.PartialResultSet{
 			Last:     true,
-			Metadata: s.ResultSet.Metadata,
+			Metadata: metadata,
 		})
 	}
 	if s.UpdateCount > 0 {
@@ -238,7 +241,7 @@ func (s *StatementResult) convertUpdateCountToResultSet(exact bool) *spannerpb.R
 	return rs
 }
 
-func (s StatementResult) getResultSetWithTransactionSet(selector *spannerpb.TransactionSelector, tx []byte) *StatementResult {
+func (s *StatementResult) getResultSetWithTransactionSet(selector *spannerpb.TransactionSelector, tx []byte) *StatementResult {
 	res := &StatementResult{
 		Type:         s.Type,
 		Err:          s.Err,
@@ -247,10 +250,16 @@ func (s StatementResult) getResultSetWithTransactionSet(selector *spannerpb.Tran
 	}
 	if s.ResultSet != nil {
 		res.ResultSet = &spannerpb.ResultSet{
-			Metadata: s.ResultSet.Metadata,
-			Rows:     s.ResultSet.Rows,
-			Stats:    s.ResultSet.Stats,
+			// Only copy the static fields of the ResultSetMetadata from the stored result to the one that will be
+			// returned for the current query. This prevents potential data races on the Transaction field.
+			Metadata: &spannerpb.ResultSetMetadata{
+				RowType:              s.ResultSet.Metadata.RowType,
+				UndeclaredParameters: s.ResultSet.Metadata.UndeclaredParameters,
+			},
+			Rows:  s.ResultSet.Rows,
+			Stats: s.ResultSet.Stats,
 		}
+
 	}
 	if _, ok := selector.GetSelector().(*spannerpb.TransactionSelector_Begin); ok {
 		if res.ResultSet == nil {
@@ -943,7 +952,7 @@ func (s *inMemSpannerServer) executeStreamingSQL(req *spannerpb.ExecuteSqlReques
 	case StatementResultError:
 		return statementResult.Err
 	case StatementResultResultSet:
-		parts, err := statementResult.ToPartialResultSets(req.ResumeToken)
+		parts, err := statementResult.ToPartialResultSets(req.ResumeToken, resWithTx.ResultSet.Metadata)
 		if err != nil {
 			return err
 		}

--- a/transaction_test.go
+++ b/transaction_test.go
@@ -160,7 +160,7 @@ func TestBeginTransactionReadOnly(t *testing.T) {
 	}
 	defer silentClose(conn)
 
-	if _, err := conn.ExecContext(ctx, "begin transaction read write"); err != nil {
+	if _, err := conn.ExecContext(ctx, "begin transaction read only"); err != nil {
 		t.Fatal(err)
 	}
 	row := conn.QueryRowContext(ctx, testutil.SelectFooFromBar, ExecOptions{DirectExecuteQuery: true})
@@ -183,11 +183,10 @@ func TestBeginTransactionReadOnly(t *testing.T) {
 	if request.GetTransaction() == nil || request.GetTransaction().GetBegin() == nil {
 		t.Fatal("missing begin transaction on ExecuteSqlRequest")
 	}
-	// TODO: Enable once transaction_read_only is picked up by the driver.
-	//readOnly := request.GetTransaction().GetBegin().GetReadOnly()
-	//if readOnly == nil {
-	//	t.Fatal("missing readOnly on ExecuteSqlRequest")
-	//}
+	readOnly := request.GetTransaction().GetBegin().GetReadOnly()
+	if readOnly == nil {
+		t.Fatal("missing readOnly on ExecuteSqlRequest")
+	}
 }
 
 func TestBeginTransactionDeferrable(t *testing.T) {

